Use a single tracer for TestProtocolClassification

brycekahle opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

Uses a single tracer for all tests within `TestProtocolClassification`. It also:
- resets the tracer client state between tests
- pauses tracepoints/kprobes/socket filters except within the `postTracerSetup` step

### Motivation

~75% of the previous time was spent attaching/detaching kprobes. 

### Additional Notes

This should be reviewed for possible side-effects between tests. We can always reset even more state within the tracer.

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

brycekahle commented on pkg/network/tracer/tracer_classification_test.go:
```diff
@@ -1314,7 +1313,7 @@ func testAMQPProtocolClassification(t *testing.T, cfg *config.Config, clientHost
 			},
 			teardown:   amqpTeardown,
-			validation: validateProtocolConnection(network.ProtocolUnknown),
+			validation: validateProtocolConnection(network.ProtocolAMQP),
```
I changed this because it seemed incorrect before, but please let me know if that isn't true.

guyarb replied:
we shouldn't classify that message type

guyarb reviewed (changes requested):
The design of the protocol classification tests meant to ensure that the tracer cannot capture traffic during the "preTracer" phase, that's to ensure the traffic we're creating during that phase, won't affect the classification of the protocol. but if the tracer is always running, than we cannot have such a guarantee.
The real problem lays with the kernel maps that we use to store the protocol of a given connection

To prove my point, I took your branch, and commented out `params.postTracerSetup(t, params.context)`, and I had many tests that passed

brycekahle commented:
Moving back to Draft while I continue to iterate on this.

brycekahle committed:
Use a single tracer for TestProtocolClassification

~75% of the previous time was spent attaching/detaching kprobes. Now we just reset the state of the tracer.

brycekahle committed:
Update ebpf-manager

brycekahle committed:
Allow small amount of time for socket filter to reattach

brycekahle committed:
only use new inner on Linux

brycekahle committed:
Fix RunDockerServer to not call t.Fail() in other goroutine

brycekahle committed:
Fix windows protocol classification inner

brycekahle committed:
Fix licenses

brycekahle committed:
Add small delay to get more reliable pause/resume behavior

brycekahle committed:
Use incorrect license because linter demands so

brycekahle commented:
@guyarb @hmahmood This is ready for another round of reviews.

guyarb commented on pkg/network/tracer/tracer_classification_linux_test.go:
```diff
@@ -0,0 +1,43 @@
+		params.preTracerSetup(t, params.context)
+	}
+	initTracerState(t, tr)
+	t.Cleanup(func() { tr.removeClient(clientID) })
```
duplication of line 21

guyarb commented on pkg/network/tracer/tracer_classification_linux_test.go:
```diff
@@ -0,0 +1,43 @@
+		params.skipCallback(t, params.context)
+	}
+	t.Cleanup(func() { tr.removeClient(clientID) })
+	t.Cleanup(func() { tr.ebpfTracer.Pause() })
```
duplication of line 40. Do we need the current call as well?

brycekahle replied:
line 40 isn't guaranteed to run if a test fails earlier. This is.

guyarb commented:
kitchen tests weren't triggered for the branch. Manually ran them

guyarb commented:
> kitchen tests weren't triggered for the branch. Manually ran them

Looks good. Approved with 2 comments for improvement/clarification

brycekahle committed:
Remove extra initTracerState call

richl-dd reviewed (approved):
👍 for WKIT-owned files
